### ISSUE TITLE
Bugs in Commander Keen 4 : Secret of the Oracle

### ISSUE THREAD

**Issue by mills32** [gerstrong/Commander-Genius#332]
First of all thanks for this awesome software!.

I'm making a mod of keen 4, using Abiathar, I tested it with this and found some "bugs".

-Links not working: sometimes a switch won't do anything (chaging tiles or removing platform blockers).
-Level entrances not working: if you create a new level entrance in abiathar it won't work on CG, you have to use the available ones in the original map.
-Some mods are not working: disabling pogo, palette changes, remove sparks from platforms, music mods (well we don't need music mods in CG).

Maybe they are just bugs caused by abiathar, (in dosbox the game works as expected).

Other features work well, like custom tile animations (really smooth), or custom EGA sprites (the hitbox calculation is less precise than in the original).

Also I wanted to ask, if it is possible to modify sprites, menu backgrounds, intro... to use more colours.

Thanks a lot!

**Comment by gerstrong** [comment]
hi @mills32 

I really would like to help on this. I'm not Abiathar user especially because Windows only. 

Can you send me your sample package which runs on Dosbox? I would let it run with CG and debug the problems.

- I wonder why level entrances do not work, because CG should do the same thing as the Keen.exe file does. But I would see it, when you send me an example like that.

- Same here with links. The algorithm is pretty much the same. So show me some examples where it does not work. I'm pretty sure, these are small unnoticed bugs, which happen not to appear in the original files.

- About some features, this is a bit more difficult. Most of the features are patched code for Exe files I guess. I have been using python scripts lately for such changes and I want to enhance this. Which features do you require. About disabling Pogo in Keen 4 I can give you a python script, with more stuff.

About modifying everything else, well. I have been enhancing the support and yes for some sprites and background there still are no ways, but if you happen to have something special, let me know and I will implement it. The hit boxes will also be improved do not worry. I'm on it.

More colors is also not a problem, images can be png, bmp, jpg even with as many colors as you want. You might have seen Keen4Plus. There are tilesets with more colors.

So let me know your thoughs, assets and whatever you want to integrate. You have my support :-)

**Comment by mills32** [comment]
Thanks. 
This is a folder containing the working MSDOS game and also with some graphics for CG.
https://www.dropbox.com/s/y5fhh88zx999bx9/Luna.7z?dl=0

The entrance level bug was solved using a new map, (maybe that was an abiathar bug).
The audio bug was also solved resetting the audio settings and now custom imf are working in CG.

Start a new game, and just go to the right on the map to enter the red and white circle to a test level.
Take the red card (gem) and the red door opens, (also a mod setting the doors to use just one tile is working fine).

1 - "remove sparks from platforms" mod is not working, as you can still see them.
2 - "Keen" still does pogo and grabs edges (I want the little robot not to do this).
3 -  After oppening the red door, activate the switch and the blue bar should become semi-transparent and not solid to let you continue, but it does nothing in CG.

So I can use python to activate mods? that's cool, I'll be needing just the 1 and 2 mods.

The palette mod is not important, as I'll be using new gfx, but I'll need to change some gfx. I saw the player sprites folder for keen4plus, so I think you could add some folders like "menu backgrounds", "loading screen", "enemies"... or something like that.

Thanks a lot.

**Comment by gerstrong** [comment]
Hi, your mod is amazing!!

I modified it a bit so it runs with CG better:

http://steinzone.de/nextcloud/index.php/s/J41XI0MWRJpZu8B

Be careful, here we have no EXE file, because it uses more python code. I have put some scripts and made the game more playable on CG.

Please use this with the preview version of CG (v.2.1.8). I'm building it right now, but SF seems to be down. So I will share another link in this issue when ready.

I like your mod very much. It is cool to see of what the Commander Keen Engine is capable of.

Keep up the good work. You have my support. About Python, I will put more features so let me know what you plan to do with your little game :-)

**Comment by gerstrong** [comment]
Here we go:

http://steinzone.de/nextcloud/index.php/s/KBt8aJhPgsg1Nqh

Try it out and let me know your thoughts

**Comment by mills32** [comment]
Switches and gem doors are working fine :).
I didn't get the python scripts working, does the game read them from the same folder?

**Comment by gerstrong** [comment]
Yes, you cannot executed them. CG does that for you. The Pogo and Cliff hanging ability should be disabled. That is one of the Python script file doings... In between other stuff

**Comment by mills32** [comment]
I'm using Windows 8 x86. python is correctly installed and working. But the py scripts are not having any effect in the game.

**Comment by gerstrong** [comment]
Can you send cglog.html ? I find that Strange. So Pogo and Cliff hanging still work? Please ensure there is no keen4e.exe in the Game path. It might Not Take all Python File if there is one...

**Comment by mills32** [comment]
Here it is: https://www.dropbox.com/s/1xdfgps4yzvta2b/CGLog.html?dl=0
The log was the same with keen4e.exe and without it.
Also Removing the exe sometimes causes error loading the maps and the program crashes.

**Comment by gerstrong** [comment]
Yes, didn't notice that for the package version python support was disabled. In my own debug version it was on. I got this fixed and am building another package:

I have extended your game package also:

https://steinzone.de/nextcloud/index.php/s/iiVAwq1NtXLmE9N

Just one script for the mad mushroom. I do not have any crashes here anymore (Tested on Windows 7). I will be sharpening some details. For example menu music. Stuff like that. Let me know what else I can do.

About the pictures... why not send me another package how you would like the file structure and I implement it.

Please use version 2.1.9 with that, not sure what older version do with it. I see sourceforge is working again. I will upload the version there and put a link here when its up.

**Comment by gerstrong** [comment]
Okay, here we go:

https://sourceforge.net/projects/clonekeenplus/files/CGenius/V2.1/

**Comment by mills32** [comment]
Thanks.

So this is not as easy as I imagined :)...

Still not working, because the program crashes. If I remove constants.py, the game runs. 
Again removing the exe made no difference.

For symplicity this is what I think it woul be, the program would look for the graphics in the gfx folder and replace the files it finds in that folder.

I added some sample stuff I whant to change, but it would read the whole contents of a graphics folder extracted with KEENGRAPH (or whatever program you used), the only difference will be the masking setup, for example KEENGRAPH uses that light green for transparency, and not the black and white image. But that's not a problem, I don't remember the other extractor, (that one produced the black and white masks).

https://www.dropbox.com/s/7x750f40bse12ni/Luna.7z?dl=0

**Comment by gerstrong** [comment]
Thanks. But still I would like to know why it crashes in your Side.Getting the right settings from exe is much harder than using Python. I will run some Tests later. Which Python did you Install? A dll is bundled. It should not be required to install afaik.constants.py disables Cliff hanging Pogo and fire particles on platforms.Am 04.03.2018 12:25 nachm. schrieb mills32 <notifications@github.com>:Thanks.
So this is not as easy as I imagined :)...
Still not working, because the program crashes. If I remove constants.py, the game runs.
Again removing the exe made no difference.
For symplicity this is what I think it woul be, the program would look for the graphics in the gfx folder and replace the files it finds in that folder.
I added some sample stuff I whant to change, but it would read the whole contents of a graphics folder extracted with KEENGRAPH (or whatever program you used), the only difference will be the masking setup, for example KEENGRAPH uses that light green for transparency, and not the black and white image. But that's not a problem, I don't remember the other extractor, (that one produced the black and white masks).
https://www.dropbox.com/s/7x750f40bse12ni/Luna.7z?dl=0

**Comment by mills32** [comment]
I have python 2.7 and 3.4.  By default all programs will use de 2.7, maybe that caused some bugs.

**Comment by gerstrong** [comment]
Well it is built against 3.5Am 04.03.2018 4:16 nachm. schrieb mills32 <notifications@github.com>:I have python 2.7 and 3.4.  By default all programs will use de 2.7, maybe that caused some bugs.

**Comment by mills32** [comment]
Now the web works, I found the forum (I posted about my mod also in the forums some time ago). and It looks like people using windows are having the same problems with pyton like this comment on the CG thread:

2.1.0 (WIndows)
 > Keen5 (.Py) Crashes when loaded, before loading bar shows up
 > Any Keen game crashes once you you get on the world, or doing the loading of the world map

I just realised I can test the game in an android phone so i'll check it :).

**Comment by mills32** [comment]
Wow sorry.. it was python.. just installed 3.50 and everything perfect, patches working.. :)

**Comment by gerstrong** [comment]
Good to know but I should check the version in future. For Android the Python Support is not ready yet. I'm working on a version where this python module and sdl 2.0 will supported... but need more time for that...Am 04.03.2018 4:51 nachm. schrieb mills32 <notifications@github.com>:Wow sorry.. it was python.. just installed 3.50 and everything perfect, patches working.. :)

**Comment by mills32** [comment]
Just one thing, How do I use flac, mod, mid etc as music? It looks like i'm doing something wrong.

Thanks!

**Comment by gerstrong** [comment]
I think it is disabled, ogg only works. But since the last days sdl Mixer is on board I will enable them for the next upload. Just need to Test some formats through... Everything sdl mixer can should be possible here also...Am 04.03.2018 10:42 nachm. schrieb mills32 <notifications@github.com>:Just one thing, How do I use flac, mod, mid etc as music? It looks like i'm doing something wrong.
Thanks!

**Comment by gerstrong** [comment]
Hi @mills32 
I have uploaded the preview version 2.1.10. I have fine tuned some stuff and added lately, especially for the Android builds which start working using SDL2 properly. There is still much to do though. 

I added support for FLAC (tested and works), mod and midi for music. 

About graphics I still have to refine the engine a bit more. When that is done much more will be possible with your little game :-)

**Comment by mills32** [comment]
Thanks! I'll test it now:)
